include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+#ifndef _IRR_BUILTIN_MATERIAL_COMPILER_GLSL_COMMON_INCLUDED_
+#define _IRR_BUILTIN_MATERIAL_COMPILER_GLSL_COMMON_INCLUDED_
+
+#include <irr/builtin/material_compiler/glsl/common_declarations.glsl>
+
+#ifndef _IRR_USER_PROVIDED_MATERIAL_COMPILER_GLSL_BACKEND_FUNCTIONS_
+	#error "You need to define 'vec3 irr_glsl_MC_getCamPos()', 'instr_t irr_glsl_MC_fetchInstr(in uint)', 'bsdf_data_t irr_glsl_MC_fetchBSDFData(in uint)' functions above"
+#endif
+
+#include <irr/builtin/glsl/math/functions.glsl>
+#include <irr/builtin/glsl/format/decode.glsl>
+
+uint instr_getNormalId(in instr_t instr)
+{
+	return (instr.y>>(INSTR_NORMAL_ID_SHIFT-32u)) & INSTR_NORMAL_ID_MASK;
+}
+uint instr_getOpcode(in instr_t instr)
+{
+	return instr.x&INSTR_OPCODE_MASK;
+}
+uint instr_getBSDFbufOffset(in instr_t instr)
+{
+	return (instr.x>>INSTR_BSDF_BUF_OFFSET_SHIFT) & INSTR_BSDF_BUF_OFFSET_MASK;
+}
+uint instr_getNDF(in instr_t instr)
+{
+	return (instr.x>>INSTR_NDF_SHIFT) & INSTR_NDF_MASK;
+}
+uint instr_getRightJump(in instr_t instr)
+{
+	return bitfieldExtract(instr.y, int(INSTR_RIGHT_JUMP_SHIFT-32u), int(INSTR_RIGHT_JUMP_WIDTH));
+}
+
+bool instr_get1stParamTexPresence(in instr_t instr)
+{
+#if defined(PARAM1_NEVER_TEX)
+	return false;
+#elif defined(PARAM1_ALWAYS_TEX)
+	return true;
+#else
+	return (instr.x&(1u<<INSTR_1ST_PARAM_TEX_SHIFT)) != 0u;
+#endif
+}
+bool instr_get2ndParamTexPresence(in instr_t instr)
+{
+#if defined(PARAM2_NEVER_TEX)
+	return false;
+#elif defined(PARAM2_ALWAYS_TEX)
+	return true;
+#else
+	return (instr.x&(1u<<INSTR_2ND_PARAM_TEX_SHIFT)) != 0u;
+#endif
+}
+bool instr_get3rdParamTexPresence(in instr_t instr)
+{
+#if defined(PARAM3_NEVER_TEX)
+	return false;
+#elif defined(PARAM3_ALWAYS_TEX)
+	return true;
+#else
+	return (instr.x&(1u<<INSTR_3RD_PARAM_TEX_SHIFT)) != 0u;
+#endif
+}
+bool instr_get4thParamTexPresence(in instr_t instr)
+{
+#if defined(PARAM4_NEVER_TEX)
+	return false;
+#elif defined(PARAM4_ALWAYS_TEX)
+	return true;
+#else
+	return (instr.x&(1u<<INSTR_4TH_PARAM_TEX_SHIFT)) != 0u;
+#endif
+}
+
+bool instr_params_getAlphaUTexPresence(in instr_t instr)
+{
+	return instr_get1stParamTexPresence(instr);
+}
+bool instr_params_getAlphaVTexPresence(in instr_t instr)
+{
+	return instr_get2ndParamTexPresence(instr);
+}
+bool instr_params_getReflectanceTexPresence(in instr_t instr)
+{
+	return instr_get4thParamTexPresence(instr);
+}
+bool instr_getSigmaATexPresence(in instr_t instr)
+{
+	return instr_get4thParamTexPresence(instr);
+}
+bool instr_getTransmittanceTexPresence(in instr_t instr)
+{
+	return instr_get1stParamTexPresence(instr);
+}
+bool instr_getWeightTexPresence(in instr_t instr)
+{
+	return instr_get1stParamTexPresence(instr);
+}
+bool instr_getOpacityTexPresence(in instr_t instr)
+{
+	return instr_get3rdParamTexPresence(instr);
+}
```

Comment:
isnt opacity (mask) implemented as a BLEND bsdf between something and the transmission BSDF? where the weight texture is the opacity texture?

Reply:
no, we rejected that idea. Long ago and dont remember why

Reply:
but opacity has not been implemented at all ?